### PR TITLE
Implement PrimaryKey for all &T where T implements PrimaryKey.

### PR DESCRIPTION
**Pull request by 0xekez**
```rust
    for<'a> &'a (K, u64): PrimaryKey<'a>,
```

This allows having type constraints like this for types that wrap storage-plus. In english: for all lifetimes `'a` `&'a (K, u64)` is a valid key. I need this generic in order to implement [`cw-future-map`](https://gist.github.com/0xekez/15fab6436ed593cbd59f0bdf7ecf1f61).

### CODE REVIEW

**Comment by 0xekez** [comment]
i am going to fork pending some resolution to how to land PRs into CosmWasm. no bad vibes, and i'd like to continue to contribute to and work with the CosmWasm repositories when there is a way to do so again.

relevant PRs:

- https://github.com/CosmWasm/cosmwasm/pull/1593
- https://github.com/cosmorama/wynddao/pull/4
- https://github.com/CosmWasm/cw-multi-test/pull/23
- https://github.com/CosmWasm/cw-storage-plus/pull/30

**Review by webmaster128 — COMMENTED** [review]
Thank you for this PR.

To be honest, I am not entirely sure if using references for primary keys is a good thing. What's the best practive for e.g. `Map<Addr>` vs. `Map<&Addr>`? You will not be able to deserialize into a key that is a reference.

@maurolacy do you know why both `impl<'a> PrimaryKey<'a> for Addr ` and `impl<'a> PrimaryKey<'a> for &'a Addr` exist and if that's desired?

**Comment by 0xekez** [comment]
@webmaster128 my understanding is that because serialization takes a reference to the type being serialized, it is more efficient to use references as keys as (1) the underlying serialization doesn't change and (2) you can avoid moves (and in turn clones) when calling methods on the map.

File: src/keys.rs [hunk 1]
```diff
@@ -134,6 +134,21 @@ impl<'a, T: PrimaryKey<'a> + Prefixer<'a> + KeyDeserialize, U: PrimaryKey<'a> +
...
 
+// implements PrimaryKey for all &T where T implements PrimaryKey.
+impl<'a, T> PrimaryKey<'a> for &'a T
+where
+    T: PrimaryKey<'a>,
+{
+    type Prefix = <T as PrimaryKey<'a>>::Prefix;
+    type SubPrefix = <T as PrimaryKey<'a>>::SubPrefix;
+    type Suffix = <T as PrimaryKey<'a>>::Suffix;
+    type SuperSuffix = <T as PrimaryKey<'a>>::SuperSuffix;
```
Comment (chipshort) [inline src/keys.rs:145]: nit: I don't think the type casts are needed, `T::Prefix` should work equally well, since `T` is unambiguously PrimaryKey.

**Review by chipshort — COMMENTED** [review]
Thanks, nice generalization.
Only change I'd require is the borrow because clippy will complain about it.

**Comment by webmaster128** [comment]
Not sure what kind of issues the CI has or had. Somehow the GitHub connection got lost. But in #31 it worked today, also for fork PRs.

When the changes requested by @chipshort are in this is good to go IMO.

**Comment by maurolacy** [comment]
> @webmaster128 my understanding is that because serialization takes a reference to the type being serialized, it is more efficient to use references as keys as (1) the underlying serialization doesn't change and (2) you can avoid moves (and in turn clones) when calling methods on the map.

This is correct. Using references is a good option for serialising stuff, while deserialization is always to an owned type, so, using references doesn't impact it.

Just seen this. Will take a detailed look and comment back.

File: src/keys.rs (hunk 1 above)
Comment (0xekez) [reply src/keys.rs]: nice! thank you.

File: src/keys.rs [hunk 2]
```diff
@@ -134,6 +134,21 @@ impl<'a, T: PrimaryKey<'a> + Prefixer<'a> + KeyDeserialize, U: PrimaryKey<'a> +
...
+
+    fn key(&self) -> Vec<Key> {
+        <T as PrimaryKey<'a>>::key(&self)
```
[the remark this answers is not in the document]
Comment (0xekez) [reply src/keys.rs]: 🙏

**Comment by 0xekez** [comment]
for some additional context, here is the map that i was able to make with these changes (we forked to include this change while the ATOM proposal was pending): [cw-wormhole](https://github.com/DA0-DA0/dao-contracts/tree/main/packages/cw-wormhole).

@chipshort - I've made your requested changes in the most recent commit. thank you!

**Review by maurolacy — APPROVED** [review]
LGTM.

File: src/keys.rs [hunk 3]
```diff
@@ -134,6 +134,21 @@ impl<'a, T: PrimaryKey<'a> + Prefixer<'a> + KeyDeserialize, U: PrimaryKey<'a> +
     }
 }
 
+// implements PrimaryKey for all &T where T implements PrimaryKey.
+impl<'a, T> PrimaryKey<'a> for &'a T
+where
+    T: PrimaryKey<'a>,
+{
+    type Prefix = <T as PrimaryKey<'a>>::Prefix;
+    type SubPrefix = <T as PrimaryKey<'a>>::SubPrefix;
```
[the remark this answers is not in the document]
Comment (maurolacy) [reply src/keys.rs:143]:
```suggestion
    type Prefix = T::Prefix;
    type SubPrefix = T::SubPrefix;
```

**Pull request merged**